Change mix calculation on `audiodelays.Echo` to allow full range.

relic-se opened the pull request:
In the current implementation of the `audiodelays.Echo.mix` parameter, the source audio sample is first combined with the echo sample and then mixed again with the source audio sample using the value of `mix` parameter. This effectively only allows you to control the level of the effect audio and not the level of the source audio (see the following graph, red=source, blue=effect).

![image](https://github.com/user-attachments/assets/778a2132-a267-4d98-aaeb-d64628771c4b)

In this update, I've extended the range of the `mix` parameter to allow you to isolate the effect audio from the source while still being able to achieve the original mixing range between 0.0 and 0.5 (see the following graph, red=source, blue=effect).

![image](https://github.com/user-attachments/assets/fb6fe588-7147-4a44-a9f3-fe268d70492a)

In my use-case, I am using CircuitPython to calculate a delay effect on an incoming signal and then mix the output using the analog pathways of my audio codec. To achieve the maximum effect, I want the digital signal to be fully "wet" (ie: `mix=1.0`) and then I can rely on the "dry" signal controlled by the codec.

gamblor21 commented:
Just to make sure I understand correctly, the idea is at 0.5 both the original and effect play at full (1.0) volume, as opposed to at 0.5 they both play at 0.5 volume?

From the graphs above it seems we lose the ability to go say 70% original 30% effect? That was the original intention of the mix parameter.

I'm not familiar enough with how other synths do this so just wondering. @todbot any thoughts if you have a moment as you originally gave me the idea?

relic-se commented:
> Just to make sure I understand correctly, the idea is at 0.5 both the original and effect play at full (1.0) volume, as opposed to at 0.5 they both play at 0.5 volume?

Because the original sample is combined with the echo sample right from the get-go, it isn't possible to get both the effect and the sample at 50% volume (without an `audiomixer`, see comments below about levels).

https://github.com/adafruit/circuitpython/blob/52b5b2da1a5443927678c953b1e8bbd3cf1cbfab/shared-module/audiodelays/Echo.c#L455

The current algorithm is essentially the following:

```
output = sample * (1 - mix) + (echo + sample) * mix
```

This can be expanded and simplified down to the following:

```
output = sample - sample * mix + echo * mix + sample * mix = sample + echo * mix
```

Essentially, you're only able to control the level of the echo and not the original source (see first graph from the top post).

> From the graphs above it seems we lose the ability to go say 70% original 30% effect? That was the original intention of the mix parameter.

I don't believe you're technically able to achieve a 70/30 mix with the current release. It would be more like 100/30 (mix=0.3). The only way you'd be able to get 70/30 exactly is by using an `audiomixer.MixerVoice` on the output. This would require `mix=0.4286` (0.3/0.7) and `level=0.7`. With this update, it would essentially be the same by the mix would be half, 0.2143.

Note: I think there might be a slight misunderstanding as to what "effect" means for `audiodelays.Echo`. I'm interpreting the "effect" as the echoed response to the original sample impulse and not the combination of that response and the sample.

Overall, I think the "mixing" paradigm needs to be implemented differently depending on the audio effect in question. For instance, `audiofilters.Filter` uses a more direct mixing algorithm which I think is appropriate for that effect (see the following graph, red=source, blue=effect, green=red + blue aka volume).

![image](https://github.com/user-attachments/assets/36ed1bfc-d33e-40a9-81d3-e7b6e8d3fbc4)

jepler commented on shared-bindings/audiodelays/Echo.c:
```diff
@@ -17,7 +17,7 @@
 
 #define DECAY_DEFAULT 0.7f
-#define MIX_DEFAULT 0.5f
+#define MIX_DEFAULT 0.25f
```
You changed this but it's not actually used below:
```
    if (mix == MP_OBJ_NULL) {
        mix = mp_obj_new_float(MICROPY_FLOAT_CONST(0.5));
}
```

relic-se replied:
I've removed those defines (as well as one in `audiofilters.Filter`) in lieu of just using the constants in the constructor. The constructor has been updated as well.

jepler reviewed (changes requested):
The documentation of the `mix` parameter is not super clear to me. Please check it over with an eye to whether it needs to be revised with this change, and whether it could be clearer.

relic-se commented:
> The documentation of the `mix` parameter is not super clear to me. Please check it over with an eye to whether it needs to be revised with this change, and whether it could be clearer.

The documentation of mix was actually fairly close. I've just added the expected result of 0.5 as to better represent the new implementation.

I noticed the description of decay was different between the constructor and the property. I found the constructor to better represent that property, so I updated the property docstring to match.

gamblor21 reviewed (commented):
I finally wrapped my head around what is going on and agree. The code looks good, I will test it on hardware soon.
Side note: maybe at some point we want to revisit all of Echo to see if we can clean it up a bit. I think as the first effect we both learned a lot inside of it.

gamblor21 reviewed (approved):
Tried it on RP2350 works as expected. Easier to understand it all by hearing it versus trying to visualize it.

relic-se commented:
@dhalbert I believe this PR is ready for merge (pending a review with write access), but would you like me to rebase it 9.2.x or keep it on 10.x (main)?

gamblor21 commented:
> @dhalbert I believe this PR is ready for merge (pending a review with write access), but would you like me to rebase it 9.2.x or keep it on 10.x (main)?

With an answer to 9.2.x or 10.x I can do a final review/merge too, to save others time.

dhalbert commented:
We don't have anything else in the 9.2.x pipeline right now, so this may not appear in a 9.2.8 that soon. If you feel this is an important bug fix, you can do `9.2.x` and it will be merged over to `main`. But otherwise `main` is fine, and you can point people to the 10.0.00 alpha and beta releases.

Note there is a failed build.

dhalbert commented:
If you want an executive decision, I'd say leave it in `main`, because we want to steer people toward trying the 10.0.0 builds.

dhalbert reviewed (approved):
I'd like to re-run the failing job, but I think this is too old to re-run. It was a transient failure, though, I believe, so never mind.

relic-se commented:
Thanks for getting this merged, @dhalbert !
